lib/Fun_Pred_Syntax.thy

```diff
@@ -0,0 +1,187 @@
+(*
+ * Copyright 2022, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+(* Syntax for using functions as predicates *)
+
+theory Fun_Pred_Syntax
+imports Main
+begin
+
+(* Functions are already instances of Boolean algebras and provide all the standard laws one
+   would like to have. Default simplifications are automatic. Search for inf/sup/uminus to
+   find further laws and/or unfold via the definitions below.
+
+   The abbreviations here provide special syntax for the function instance of Boolean
+   algebras only, leaving other instances (such as "bool") untouched. *)
```

Comment:
I'll be honest, I don't really follow what is happening here and the comment doesn't help me that much either, likely due to `boolean_algebra` being a new thing for me.
Reading the header of this theory, "Syntax for using functions as predicates", it would imply that any functions can be used as predicates?
You're excluding "bool". Is that to avoid being able to say "True and True" thus duplicating the normal logical operators?

Reply:
> You're excluding "bool". Is that to avoid being able to say "True and True" thus duplicating the normal logical operators?

Yes, precisely, I don't want people to start using these for `True and True` for bool, just because it's theoretically possible.

> Reading the header of this theory, "Syntax for using functions as predicates", it would imply that any functions can be used as predicates?

Any predicate-like function type, yes. I.e. `'s => bool` as before or any `'a => 'b => 'c => 'd => .. => bool` (where and of the type variables can themselves be functions or anything else) or any `'x => 'y` where `'y` is a boolean algebra. As long as it ends in bool or a in a boolean algebra, it'll work.

Basically, `'s => bool` is what we'd usually call predicates (functions from something to bool), and this generalises it to any number of arguments where the arguments can be anything and the result doesn't have to be strictly bool, but is bool-like enough to satisfy the laws of booleans (= boolean_algebra).

Reply:
Would it make any sense to update the header to reflect that a bit better, if this theory provides syntax for predicate-like functions with an arbitrary number of arguments? I'm ok either way, but speaking from a perspective of having been confused after reading it.

Reply:
Yup, can add a bit more to the header.

Reply:
Thumbs up.